### PR TITLE
added documentation for recursion-depth

### PR DESCRIPTION
**Pull request by Richajaishwal0**
### Description
Recursion depth management ensures that recursive functions do not exceed stack limits, avoiding errors like stack overflow. Backtracking is an approach that uses recursion to explore all potential solutions to a problem, incrementally building candidates and abandoning those that do not meet constraints. In programs, these concepts often include solving puzzles (like Sudoku), combinatorial problems (like generating permutations), and optimization tasks (like the N-Queens problem), effectively allowing for thorough exploration of solution spaces while managing memory and performance.

Fixes # 1702

### Type of change

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Documentation update

### Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: docs/Recursion-depths/advanced-topics.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
+
+---
+Advanced topics in recursion, including dynamic programming and recursion depth in complex algorithms.
+| ID                      | Title                         | Sidebar Label    | Sidebar Position | Description                                            | Tags                     |
+|-------------------------|-------------------------------|------------------|------------------|--------------------------------------------------------|--------------------------|
+| recursion-depth-advanced | Advanced Recursion Depth Topics| Advanced Topics   | 8                | Advanced concepts related to recursion depth, including dynamic programming and backtracking. | advanced, recursion, algorithms |
```
Comment (ajay-dhangar) [inline docs/Recursion-depths/advanced-topics.md]: please follow our docs format and style

File: docs/Recursion-depths/basic-concepts.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
+
+| ID                         | Title                             | Sidebar Label    | Sidebar Position | Description                                            | Tags                      |
+|----------------------------|-----------------------------------|------------------|------------------|--------------------------------------------------------|---------------------------|
+| recursion-depth-basic-concepts | Basic Concepts of Recursion Depth | Basic Concepts    | 2                | Introduction to recursion and understanding recursion depth. | recursion, depth, basics  |
```
Comment (ajay-dhangar) [inline docs/Recursion-depths/basic-concepts.md]: change format

File: docs/Recursion-depths/examples/binary-search-depth.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
+
+| ID                          | Title                        | Sidebar Label          | Sidebar Position | Description                                         | Tags                           |
+|-----------------------------|------------------------------|------------------------|------------------|-----------------------------------------------------|--------------------------------|
+| binary-search-recursion-depth| Binary Search Recursion Depth | Binary Search Example   | 5                | An example of recursion depth with binary search.   | binary-search, recursion, depth |
```
Comment (ajay-dhangar) [inline docs/Recursion-depths/examples/binary-search-depth.md]: ❌

File: docs/Recursion-depths/examples/binary-search-depth.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+3. **Optimization**:
+   - Knowing recursion depth helps optimize the recursive function for a balance between efficiency and resource use, especially when working with large datasets.
+
+## Practical Example of Recursion Depth
+
+- For an array of size 8, the maximum recursion depth is:
+  \[
+  \log_2(8) = 3
+  \]
```
Comment (ajay-dhangar) [inline docs/Recursion-depths/examples/binary-search-depth.md]: use `$$ ... $$`

File: docs/Recursion-depths/examples/binary-search-depth.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+- For an array of size 1024, the maximum depth is:
+  \[
+  \log_2(1024) = 10
+  \]
```
Comment (ajay-dhangar) [inline docs/Recursion-depths/examples/binary-search-depth.md]: use `$$ ... $$` for math equation

File: docs/Recursion-depths/examples/factorial-depth.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
+
+---
+Detailed example of recursion depth in the factorial function.
+
+
+| ID                        | Title                       | Sidebar Label         | Sidebar Position | Description                                             | Tags                           |
+|---------------------------|-----------------------------|-----------------------|------------------|---------------------------------------------------------|--------------------------------|
+| factorial-recursion-depth  | Factorial Recursion Depth   | Factorial Example      | 3                | An example of recursion depth with factorial calculations. | factorial, recursion, depth    |
+
```
Comment (ajay-dhangar) [inline docs/Recursion-depths/examples/factorial-depth.md]: ❌

File: docs/Recursion-depths/examples/factorial-depth.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+## Example Code
+```python
+def factorial(n):
+    if n == 0 or n == 1:
+        return 1
+    return n * factorial(n - 1)
+
+### Explanation of the Code
+
+- **Base Case**: The function returns `1` when \( n \) is 0 or 1, which stops the recursion.
```
Comment (ajay-dhangar) [inline docs/Recursion-depths/examples/factorial-depth.md]: replace `\(n \)` with `$n$`

File: docs/Recursion-depths/examples/factorial-depth.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+- **Recursive Calculation**: Each call to `factorial(n)` results in a call to `factorial(n - 1)`, building up the multiplication as the stack unwinds.
+
+### Importance of Recursion Depth in Factorial Calculations
+
+Understanding recursion depth in the factorial function is important because:
+
+- **Efficiency**: The maximum recursion depth equals the input \( n \), leading to \( O(n) \) time complexity.
+- **Memory Consumption**: Each recursive call adds to the stack, so for large \( n \), the program may run into stack overflow issues.
```
Comment (ajay-dhangar) [inline docs/Recursion-depths/examples/factorial-depth.md]: Please apply for the math equation `$` for inline and `$$` for block line. These are similar to \( or \) and \[ or \].

File: docs/Recursion-depths/examples/factorial-depth.md [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+- **Optimization**: Using an iterative approach can prevent deep recursion.
+
+### Practical Example of Recursion Depth
+
+For an input value of 5, the maximum recursion depth is:
+
+\[
+\text{depth} = 5
+\]
```
Comment (ajay-dhangar) [inline docs/Recursion-depths/examples/factorial-depth.md]: use `$$ ... $$`

File: docs/Recursion-depths/examples/fibonacci-depth.md [hunk 10]
```diff
@@ -0,0 +1,48 @@
+
+---
+Detailed example of recursion depth in the Fibonacci sequence.
+
+
+| ID                        | Title                      | Sidebar Label         | Sidebar Position | Description                                                  | Tags                           |
+|---------------------------|----------------------------|-----------------------|------------------|--------------------------------------------------------------|--------------------------------|
+| fibonacci-recursion-depth  | Fibonacci Recursion Depth   | Fibonacci Example      | 4                | An example of recursion depth with Fibonacci sequence calculations. | fibonacci, recursion, depth    |
+
```
Comment (ajay-dhangar) [inline docs/Recursion-depths/examples/fibonacci-depth.md]: ❌

File: docs/Recursion-depths/examples/fibonacci-depth.md [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+        return 0
+    elif n == 1:
+        return 1
+    return fibonacci(n - 1) + fibonacci(n - 2)
+
+### Explanation of the Code
+
+- **Base Case**: The function returns `0` if \( n \) is 0, and returns `1` if \( n \) is 1. These cases stop the recursion.
+- **Recursive Calculation**: The function calls itself twice for \( n-1 \) and \( n-2 \), leading to exponential growth in calls.
+
```
Comment (ajay-dhangar) [inline docs/Recursion-depths/examples/fibonacci-depth.md]: use `$`

File: docs/Recursion-depths/examples/fibonacci-depth.md [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+### Importance of Recursion Depth in Fibonacci Calculations
+
+Understanding recursion depth in Fibonacci calculations is crucial because:
+
+- **Efficiency**: The naive recursive Fibonacci approach has an exponential time complexity of \( O(2^n) \).
+- **Memory Consumption**: Each recursive call contributes to the stack, leading to potential overflow for large \( n \).
+- **Optimization**: Memoization can greatly improve performance by storing previously computed values.
```
Comment (ajay-dhangar) [inline docs/Recursion-depths/examples/fibonacci-depth.md]: use `$`

File: docs/Recursion-depths/examples/fibonacci-depth.md [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+### Practical Example of Recursion Depth
+
+For an input value of 5, the recursion depth will be:
+
+\[
+\text{depth} = 5 \quad (\text{maximum depth in calls})
+\]
+
```
Comment (ajay-dhangar) [inline docs/Recursion-depths/examples/fibonacci-depth.md]: use `$$`

File: docs/Recursion-depths/handling-depth-errors.md [hunk 14]
```diff
@@ -0,0 +1,54 @@
+| ID                        | Title                          | Sidebar Label | Sidebar Position | Description                                                   | Tags                          |
+|---------------------------|--------------------------------|---------------|------------------|---------------------------------------------------------------|-------------------------------|
+| recursion-depth-errors    | Handling Recursion Depth Errors | Depth Errors   | 7                | Handling stack overflow and other errors due to high recursion depth. | errors, recursion, stack-overflow |
+
```
Comment (ajay-dhangar) [inline docs/Recursion-depths/handling-depth-errors.md]: ❌

File: docs/Recursion-depths/handling-depth-errors.md [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+Here’s a simple recursive function that calculates the factorial of a number:
+
+```python
+def factorial(n):
+    if n < 0:  # Error check for negative input
+        return "Error: Negative input"
+    elif n == 0 or n == 1:
+        return 1
+    return n * factorial(n - 1)
+
```
Comment (ajay-dhangar) [inline docs/Recursion-depths/handling-depth-errors.md]: close your code example with ```` ``` ````

File: docs/Recursion-depths/performance-consideration.md [hunk 16]
```diff
@@ -0,0 +1,19 @@
+
+Discussion on how recursion depth impacts performance, with tips to optimize.
+
+
+| ID                        | Title                           | Sidebar Label            | Sidebar Position | Description                                              | Tags                           |
+|---------------------------|---------------------------------|--------------------------|------------------|----------------------------------------------------------|--------------------------------|
+| recursion-depth-performance| Recursion Depth and Performance  | Performance Considerations | 6                | Understanding how recursion depth affects algorithm performance. | performance, recursion, optimization
```
Comment (ajay-dhangar) [inline docs/Recursion-depths/performance-consideration.md]: ❌

**Review by ajay-dhangar — CHANGES_REQUESTED** [review]
Please follow our format and syntax when adding documentation. Refer to our docs format guidelines and our contributions guidelines.

File: docs/Recursion-depths/Recursion.md [hunk 17]
```diff
@@ -0,0 +1,65 @@
+| id                         | sidebar_position | title                     | sidebar_label               | tags                 |
+|----------------------------|------------------|---------------------------|-----------------------------|----------------------|
+| recursion-depth-overview   | 1                | Recursion Depth Overview  | Overview                    | recursion, algorithms, depth |
+
```
Comment (ajay-dhangar) [inline docs/Recursion-depths/Recursion.md]: this is not a good way so, please follow our format

**Comment by Richajaishwal0** [comment]
@ajay-dhangar I have updated the documentation (#1702). Please check and let me know if anything needs to be modified.

**Review by ajay-dhangar — APPROVED** [review]
Done

**Pull request merged**